ledger/simulation/simulator_test.go

```diff
@@ -0,0 +1,662 @@
+	"github.com/algorand/go-algorand/protocol"
+	"github.com/algorand/go-algorand/test/partitiontest"
+	"github.com/stretchr/testify/require"
+)
+
+// ==============================
+// > Simulation Test Helpers
+// ==============================
+
+type account struct {
+	addr     basics.Address
+	sk       *crypto.SignatureSecrets
+	acctData basics.AccountData
+}
+
+func prepareSimulatorTest(t *testing.T) (l *data.Ledger, accounts []account, makeTxnHeader func(sender basics.Address) transactions.Header) {
+	genesisInitState, keys := ledgertesting.GenerateInitState(t, protocol.ConsensusCurrentVersion, 100)
+
+	// Prepare ledger
+	const inMem = true
+	cfg := config.GetDefaultLocal()
+	cfg.Archival = true
+	log := logging.TestingLog(t)
+	log.SetLevel(logging.Warn)
+	realLedger, err := ledger.OpenLedger(log, t.Name(), inMem, genesisInitState, cfg)
+	require.NoError(t, err, "could not open ledger")
+
+	l = &data.Ledger{Ledger: realLedger}
+	require.NotNil(t, l)
+
+	// Reformat accounts
+	accounts = make([]account, len(keys)-2) // -2 for pool and sink accounts
+	i := 0
+	for addr, key := range keys {
+		if addr == ledgertesting.PoolAddr() || addr == ledgertesting.SinkAddr() {
+			continue
+		}
+
+		acctData := genesisInitState.Accounts[addr]
+		accounts[i] = account{addr, key, acctData}
+		i++
+	}
+
+	// txn header generator
+	hdr, err := l.BlockHdr(l.Latest())
+	require.NoError(t, err)
+	makeTxnHeader = func(sender basics.Address) transactions.Header {
+		return transactions.Header{
+			Fee:         basics.MicroAlgos{Raw: 1000},
+			FirstValid:  hdr.Round,
+			GenesisID:   hdr.GenesisID,
+			GenesisHash: hdr.GenesisHash,
+			LastValid:   hdr.Round + basics.Round(1000),
+			Note:        []byte{240, 134, 38, 55, 197, 14, 142, 132},
+			Sender:      sender,
+		}
+	}
+
+	return
+}
+
+func makeTestClient() libgoal.Client {
+	c, err := libgoal.MakeClientFromConfig(libgoal.ClientConfig{
+		AlgodDataDir: "NO_DIR",
+	}, libgoal.DynamicClient)
+	if err != nil {
+		panic(err)
+	}
+
+	return c
+}
+
+// Attach group ID to a transaction group. Mutates the group directly.
+func attachGroupID(txgroup []transactions.SignedTxn) error {
+	txnArray := make([]transactions.Transaction, len(txgroup))
+	for i, txn := range txgroup {
+		txnArray[i] = txn.Txn
+	}
+
+	client := makeTestClient()
+	groupID, err := client.GroupID(txnArray)
+	if err != nil {
+		return err
+	}
+
+	for i := range txgroup {
+		txgroup[i].Txn.Header.Group = groupID
+	}
+
+	return nil
+}
+
+func uint64ToBytes(num uint64) []byte {
+	ibytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(ibytes, num)
+	return ibytes
+}
+
+// ==============================
+// > Sanity Tests
+// ==============================
+
+// We want to be careful that the Algod ledger does not move on to another round
+// so we confirm here that all ledger methods which implicitly access the current round
+// are overriden within the `simulatorLedger`.
+func TestNonOverridenDataLedgerMethodsUseRoundParamter(t *testing.T) {
```

Comment:
@jdtzmn Minor typo in _paramter_.  

```suggestion
func TestNonOverridenDataLedgerMethodsUseRoundParameter(t *testing.T) {
```